### PR TITLE
refactor(breaking): Consistent API surface for devices

### LINKED ISSUE #17: Traits for devices
In PROS 4, all smart devices have a common `is_installed` method that can check the device's connection status as well as a `get_port` method that returns the device's port index (https://github.com/purduesigbots/pros/pull/539). This is likely best going to be expressed through a common `SmartDevice` trait once that comes around.

In addition, some sensors share mostly common behaviors:
- The IMU, GPS, and AdiGyro devices all share similar methods for reading orientation. (VEXCode has a `vex::guido` interface for this purpose).
- The potentiometer, encoders (including IMEs), and rotation sensors all report some degree of angular position. (Okapi abastracts over these using a `ContinuousRotarySensor` class... bit of a mouthful but it works)

Might also be nice to have traits for these too, since it'd make writing generic code for common device functionality a whole lot easier.

### PR DESCRIPTION
**Pull request by Tropix126**
Blocked by #55 (will merge from upstream and make necessary changes once available)
Partially resolves #17

## Describe the changes this PR makes. Why should it be merged?
Fixes some consistency issues in the current smart device/sensor implementations and refactors file structure for better organization.
- Derives `Debug, Eq, PartialEq` for all device types.
- Attempts to derive `Default, Debug, Clone, Copy, Eq, PartialEq` for all device-related data structs if possible, unless `Default` doesn't make any sense (such as with bitflags).
- Makes all methods that modify device state take `&mut self`.
- `SmartPort` struct for representing a single smart port. Unsafe to construct for now, until peripherals becomes a thing.
- `SmartDevice` trait for all smart devices for checking device type, connected status.
- Refactored `AdiPort` to support expander modules.

## Additional Context
- These are (significantly) breaking changes (semver: major).

### CODE REVIEW

**Review by Gavin-Niederman — CHANGES_REQUESTED** [review]
Just a few things I saw immediately. This is a lot of code that got changed so I probably didn't catch everything.

File: packages/pros/src/devices/adi/mod.rs [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    /// Ports are indexed starting from 1.
+    pub fn index(&self) -> u8 {
+        self.index
+    }
+
+    /// Get the index of this port's associated [`AdiExpander`] smart port, or `None` if this port is not
+    /// associated with an expander.
+    pub fn expander_index(&self) -> Option<u8> {
+        self.expander_index
+    }
```
[the remark this answers is not in the document]
Comment (Tropix126) [reply packages/pros/src/devices/adi/mod.rs:64]: I don't think it should be (it's something I thought about), even though there are a few common behaviors between devices and ports.
- We fundamentally treat them differently. An `AdiDevice` or `SmartDevice` takes ownership of a port, not another device really. Both exist at the same time and it's a little strange imo to treat them both as a registered device and unregistered device on the same index.
- On `AdiDevice`s, there's no real `AdiDeviceType` available to match an unregistered port (`SmartDeviceType` is a different story, though).

File: packages/pros/src/devices/adi/digital.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                PROS_ERR,
+                pros_sys::ext_adi_digital_get_new_press(
+                    self.port.internal_expander_index(),
+                    self.port.index()
+                )
+            ) != 0
+        })
+    }
+
+    /// Gets the current value of a digital input pin.
```
Comment (Gavin-Niederman) [inline packages/pros/src/devices/adi/digital.rs:30]: These docs are exactly the same, but the functions do different things.

File: packages/pros/src/devices/adi/encoder.rs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct AdiEncoder {
+    raw: ext_adi_encoder_t,
+    port_top: AdiPort,
+    port_bottom: AdiPort,
+}
+
+impl AdiEncoder {
+    /// Create an AdiEncoder, returning err `AdiError::InvalidPort` if the port is invalid.
```
Comment (Gavin-Niederman) [inline packages/pros/src/devices/adi/encoder.rs:14]: The port cannot be invalid.

File: packages/pros/src/devices/adi/gyro.rs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    /// Gets the current gyro angle in tenths of a degree. Unless a multiplier is applied to the gyro, the return value will be a whole number representing the number of degrees of rotation times 10.
+    ///
+    /// There are 360 degrees in a circle, thus the gyro will return 3600 for one whole rotation.
+    pub fn value(&self) -> Result<f64, AdiError> {
+        Ok(bail_on!(PROS_ERR.into(), unsafe {
+            pros_sys::ext_adi_gyro_get(self.raw)
+        }))
+    }
+
+    /// Gets the current gyro angle in tenths of a degree. Unless a multiplier is applied to the gyro, the return value will be a whole number representing the number of degrees of rotation times 10.
```
Comment (Gavin-Niederman) [inline packages/pros/src/devices/adi/gyro.rs:31]: This doc comment is wrong.

**Review by Gavin-Niederman — CHANGES_REQUESTED** [review]
Just some small things. After this I'm fairly sure that we can get this merged.

**Review by Gavin-Niederman — APPROVED** [review]
LGTM (Legacy Getters for Those Motors)

**Pull request merged**